presto-main/src/main/java/com/facebook/presto/operator/aggregation/DecimalAverageAggregation.java

```diff
@@ -0,0 +1,192 @@
+import static com.facebook.presto.operator.aggregation.AggregationUtils.generateAggregationName;
+import static com.facebook.presto.spi.type.LongDecimalType.unscaledValueToBigInteger;
+import static com.facebook.presto.spi.type.StandardTypes.DECIMAL;
+import static com.facebook.presto.util.Reflection.methodHandle;
+import static java.math.BigDecimal.ROUND_DOWN;
+
+public class DecimalAverageAggregation
+        extends SqlAggregationFunction
+{
+    public static final DecimalAverageAggregation DECIMAL_AVERAGE_AGGREGATION = new DecimalAverageAggregation();
+    private static final String NAME = "avg";
+    private static final MethodHandle SHORT_DECIMAL_INPUT_FUNCTION = methodHandle(DecimalAverageAggregation.class, "inputShortDecimal", Type.class, BigIntegerAndLongState.class, Block.class, int.class);
+    private static final MethodHandle LONG_DECIMAL_INPUT_FUNCTION = methodHandle(DecimalAverageAggregation.class, "inputLongDecimal", Type.class, BigIntegerAndLongState.class, Block.class, int.class);
+
+    private static final MethodHandle SHORT_DECIMAL_OUTPUT_FUNCTION = methodHandle(DecimalAverageAggregation.class, "outputShortDecimal", ShortDecimalType.class, BigIntegerAndLongState.class, BlockBuilder.class);
+    private static final MethodHandle LONG_DECIMAL_OUTPUT_FUNCTION = methodHandle(DecimalAverageAggregation.class, "outputLongDecimal", LongDecimalType.class, BigIntegerAndLongState.class, BlockBuilder.class);
+
+    private static final MethodHandle COMBINE_FUNCTION = methodHandle(DecimalAverageAggregation.class, "combine", BigIntegerAndLongState.class, BigIntegerAndLongState.class);
+
+    public DecimalAverageAggregation()
+    {
+        super(NAME, ImmutableList.of(Signature.withVariadicBound("T", DECIMAL)), "T", ImmutableList.of("T"));
+    }
+
+    @Override
+    public String getDescription()
+    {
+        return "Calculates the average value";
+    }
+
+    @Override
+    public InternalAggregationFunction specialize(Map<String, Type> types, List<TypeSignature> parameterTypes, TypeManager typeManager, FunctionRegistry functionRegistry)
+    {
+        Type type = types.get("T");
+        return generateAggregation(type);
+    }
+
+    private static InternalAggregationFunction generateAggregation(Type type)
+    {
+        DynamicClassLoader classLoader = new DynamicClassLoader(DecimalAverageAggregation.class.getClassLoader());
+
+        List<Type> inputTypes = ImmutableList.of(type);
+
+        MethodHandle inputFunction;
+        MethodHandle combineFunction;
+        MethodHandle outputFunction = null;
+        Class<? extends AccumulatorState> stateInterface = BigIntegerAndLongState.class;
+        AccumulatorStateSerializer<?> stateSerializer = new BigIntegerAndLongStateSerializer();
+
+        if (type.getJavaType() == long.class) {
+            inputFunction = SHORT_DECIMAL_INPUT_FUNCTION;
+            outputFunction = SHORT_DECIMAL_OUTPUT_FUNCTION;
+            combineFunction = COMBINE_FUNCTION;
```

Comment:
I removed the whole combineFunction variable, assigning using the static field directly